[ENH] enhance naive forecaster with new strategies

TinaJin0228 opened the pull request:
<!--
Thanks for contributing a pull request! Please ensure you have taken a look at our
contribution guide: https://www.aeon-toolkit.org/en/latest/contributing.html.

Feel free to delete sections of this template if they do not apply to your PR,
avoid submitting a blank template or empty sections.
If you are a new contributor, do not delete this template without a suitable
replacement or reason. If in doubt, ask for help. We're here to help!

Please be aware that we are a team of volunteers so patience is
necessary when waiting for a review or reply. There may not be a quick turnaround for
reviews during slow periods. While we value all contributions big or small, pull
requests which do not follow our guidelines may be closed.
-->

#### Reference Issues/PRs

<!--
Example: Fixes #1234. See also #3456.
Please use keywords (e.g., Fixes) to create link to the issues or pull requests
you resolved, so that they will automatically be closed when your pull request
is merged. See https://github.com/blog/1506-closing-issues-via-pull-requests
-->

Fixes #2823. 


#### What does this implement/fix? Explain your changes.
1.  New strategies for the naive forecaster, including mean and seasonal last, are added. 
2. test_naive.py for testing the naive forecaster is added. 


<!--
A clear and concise description of what you have implemented.
-->

#### Does your contribution introduce a new dependency? If yes, which one?
No. 

<!--
If your contribution does add a dependency, we may suggest adding it as an
optional/soft dependency to keep external dependencies of the core aeon package
to a minimum.
-->

#### Any other comments?

<!--
Any other information that is important to this PR or helpful for reviewers.
-->

### PR checklist

<!--
Please go through the checklist below. Please feel free to remove points if they are
not applicable. To check a box, replace the space inside the square brackets with an
'x' i.e. [x].
-->

##### For all contributions
- [x] I've added myself to the [list of contributors](https://github.com/aeon-toolkit/aeon/blob/main/.all-contributorsrc). Alternatively, you can use the [@all-contributors](https://allcontributors.org/docs/en/bot/usage) bot to do this for you **after** the PR has been merged.
- [x] The PR title starts with either [ENH], [MNT], [DOC], [BUG], [REF], [DEP] or [GOV] indicating whether the PR topic is related to enhancement, maintenance, documentation, bugs, refactoring, deprecation or governance.

##### For new estimators and functions
- [x] I've added the estimator/function to the online [API documentation](https://www.aeon-toolkit.org/en/latest/api_reference.html).
- [ ] (OPTIONAL) I've added myself as a `__maintainer__` at the top of relevant files and want to be contacted regarding its maintenance. Unmaintained files may be removed. This is for the full file, and you should not add yourself if you are just making minor changes or do not want to help maintain its contents.

##### For developers with write access
- [ ] (OPTIONAL) I've updated aeon's [CODEOWNERS](https://github.com/aeon-toolkit/aeon/blob/main/CODEOWNERS) to receive notifications about future changes to these files.


<!--
Thanks for contributing!
-->

TonyBagnall commented on aeon/forecasting/_naive.py:
```diff
@@ -1,27 +1,91 @@
+                predictions[i] = self._fitted_last_season_[i % m]
+        return predictions
 
     def _forecast(self, y, exog=None):
```
you dont need this, because it is doing the same as the base class forecast, can just lose the whole funtion in this class

TonyBagnall commented on aeon/forecasting/_naive.py:
```diff
@@ -1,27 +1,91 @@
+            )
         return self
 
     def _predict(self, y=None, exog=None):
```
predict should just return a single value, not all the values for the horizon

MatthewMiddlehurst replied:
This should really be a general test 🙂

TinaJin0228 commented:
I have made these modifications in the new commit:
1. write documentation, both API page and examples page
2. delete redundant `_forecast` function, modify `_predict` to return a single value (sorry for my carelessness)
3. modify docstring and test file according to reviews

TonyBagnall reviewed (approved):
LGTM, I will let matthew see if his comments are resolved, but think they are

MatthewMiddlehurst commented on aeon/forecasting/_naive.py:
```diff
@@ -9,39 +9,34 @@ class NaiveForecaster(BaseForecaster):
-        horizon : int, default=1
-            The number of time steps ahead to forecast.
-        """
+        """Initialize NaiveForecaster."""
```
Don't need this comment

TinaJin0228 replied:
got it

MatthewMiddlehurst commented on aeon/forecasting/_naive.py:
```diff
@@ -9,39 +9,34 @@ class NaiveForecaster(BaseForecaster):
 
     def _fit(self, y, exog=None):
         """Fit Naive forecaster to training data `y`."""
+        self._fitted_scalar_value_ = None  # For 'last' and 'mean' strategies
+        self._fitted_last_season_ = np.array([])  # For 'seasonal_last', init as empty
```
these do not need to end in _ if they start with _. Also it is probably better to not initialise them here and only create them if they are relevant IMO but that may not be best practice.

TinaJin0228 replied:
I agree with you; I have deleted the initialization.

MatthewMiddlehurst commented on aeon/forecasting/_naive.py:
```diff
@@ -58,34 +53,12 @@ def _fit(self, y, exog=None):
-                "Valid strategies are 'last', 'mean', 'seasonal_last'."
-            )
-        return predictions
+    # The class will inherit the _forecast method from BaseForecaster.
```
Dont need this comment, it is implied by the lack of the function.

MatthewMiddlehurst commented on aeon/forecasting/tests/test_naive.py:
```diff
@@ -34,12 +31,13 @@ def test_naive_forecaster_seasonal_last_strategy():
     predictions = forecaster.predict()
-    expected = np.array([6, 7, 8, 6])  # Repeats 6, 7, 8, then 6
+    expected = np.array(6)
     np.testing.assert_array_equal(predictions, expected)
```
@TonyBagnall probably relevant to some of the discussions going on elsewhere

TonyBagnall commented on aeon/forecasting/tests/test_naive.py:
```diff
@@ -34,12 +31,13 @@ def test_naive_forecaster_seasonal_last_strategy():
 
 
-def test_naive_forecaster_forecast_method(sample_data):
+def test_naive_forecaster_forecast_method():
```
its gone

TonyBagnall reviewed (approved):
it would be good to parameterise the test, but that could be a good first issue for someone else :) Resolve the conflict in notebooks by using yours, then imo it can go in

TinaJin0228 commented:
> Please avoid force pushing if possible, it mangles the commit history which makes it difficult to compare changes since previous reviews were made.

got it

TinaJin0228 commented on aeon/forecasting/_naive.py:
```diff
@@ -1,27 +1,57 @@
     def _predict(self, y=None, exog=None):
-        """Predict using Naive forecaster."""
-        return self.last_value_
+        """Predict a single value self.horizon steps ahead."""
+        if self.strategy == "last" or self.strategy == "mean":
+            return self._fitted_scalar_value
```
Currently I believe the code in _predict should be modified to support "y!= None" cases. But yes we can also add a tag indicating that this forecaster does not support input series in predict...
Btw what do you mean by "but trying to avoid that here."?

MatthewMiddlehurst replied:
> "but trying to avoid that here."

Rather talk about what to do in this case than framework discussion. The framework discussion is valid but it would be good to get this in next week for the next release.

So either a) accept y or b) raise an exception really. Only relevant for the seasonal one, but i would assume you determine which of the 4 values (assuming a period of 4) to return using the length of the y input? so incorporate len(y) % self.seasonal_period then do the horizon from there or something

hadifawaz1999 reviewed (approved):
looks good to me but it seems the horizon parameter is missing from docstring?

TinaJin0228 commented:
> looks good to me but it seems the horizon parameter is missing from docstring?

Since it's already in the BaseForecaster's docstring, I thought it wasn’t necessary to repeat it here?

MatthewMiddlehurst commented:
yes we need to document that here as well, classes do not inherit docstrings
